src/interfaces/chain.cpp

```diff
@@ -358,22 +359,38 @@ class ChainImpl : public Chain
     {
         ::uiInterface.ShowProgress(title, progress, resume_possible);
     }
-    std::unique_ptr<Handler> handleNotifications(Notifications& notifications) override
+    std::unique_ptr<Handler> handleNotifications(Notifications& notifications, SnapshotFn snapshot_fn) override
     {
-        return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
+        AssertLockNotHeld(::cs_main);
+        AssertLockNotHeld(::mempool.cs);
+
+        // Declare packaged task to run in notification thread, and to first
+        // send the snapshot before enabling subsequent notifications.
+        std::vector<CTransactionRef> snapshot;
+        std::packaged_task<std::unique_ptr<Handler>()> task([&] {
+            if (snapshot_fn) snapshot_fn(std::move(snapshot));
+            return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
```

Comment:
Mempool can change after the snapshot but before registering the interface and so those changes will be missed.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271292127

> Mempool can change after the snapshot but before registering the interface and so those changes will be missed.

This isn't true, and the comment below, "Hold locks while scheduling the task so notifications about added and removed transactions after the snapshot arrive after the snapshot" is specifically referring to this case.

Let me know if I should make it more clear, but as long as the locks are held, transactions can't be added / removed by other threads, and as long as CallFunctionInValidationInterfaceQueue is called before transactions are added/removed, the handler will be registered before the queued notifications signal it.

Reply:
But the queue can be non empty when CallFunctionInValidationInterfaceQueue is called?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271375746

> But the queue can be non empty when CallFunctionInValidationInterfaceQueue is called?

I think you are referring to a different problem now (problem of old notifications being received when they shouldn't be, which is different from the problem of new notifications not being received when they should be). The problem of old notifications being received when they shouldn't be is what happens without this PR in the current code:

https://github.com/bitcoin/bitcoin/blob/2c364fde423e74b4e03ebcff4582a9db7a6c4e4b/src/interfaces/chain.h#L273-L281

The problem is fixed by this PR by not constructing NotificationsHandlerImpl until after snapshot_fn is called. The comment there "to first send the snapshot before enabling subsequent notifications" is about this.

Reply:
`AddToProcessQueue` tricked me, what really goes to that queue are notifications.